File: hackerrank/problem solving/data structures/1 - arrays/5 - sparse arrays/sparse_arrays.py
# https://www.hackerrank.com/challenges/sparse-arrays/problem?isFullScreen=true

from collections import Counter


def main():
    n = int(input())
    strings = read_lines(n)
    strings_to_counts = Counter(strings)

    n = int(input())
    queries = read_lines(n)

    print(*counts_from(strings_to_counts, queries), sep='\n')


def read_lines(n):
    return (input() for _ in range(n))


# n: length of queries
# max_query_string_length: 20
# T: O(n * max_query_string_length) = O(n * 20) = O(n)
# S: O(1) extra space
def counts_from(freq_map, queries):
    return (freq_map[q] for q in queries)


if __name__ == '__main__':
    main()

File: hackerrank/problem solving/data structures/1 - arrays/5 - sparse arrays/sparse_arrays.cs
// https://www.hackerrank.com/challenges/sparse-arrays/problem?isFullScreen=true

int n = int.Parse(Console.ReadLine()!);
string[] strings = ReadLines(n);
Dictionary<string, int> stringsToCounts = Counter(strings);

n = int.Parse(Console.ReadLine()!);
string[] queries = ReadLines(n);

foreach (var x in CountsFrom(stringsToCounts, queries))
    Console.WriteLine(x);

string[] ReadLines(int n) => Enumerable.Range(0, n).Select(_ => Console.ReadLine()!).ToArray();

Dictionary<string, int> Counter(string[] arr) => arr.GroupBy(x => x).ToDictionary(g => g.Key, g => g.Count());

// n: length of queries
// maxQueryStringLength: 20
// T: O(n * maxQueryStringLength) = O(n * 20) = O(n)
// S: O(n) extra space
int[] CountsFrom(Dictionary<string, int> freqMap, string[] queries)
{
    return queries.Select(q => freqMap.GetValueOrDefault(q)).ToArray();
}
